edb/server/pgcon/rust_transport.py

```diff
@@ -0,0 +1,475 @@
+    ConnectionParams,
+    SSLMode,
+    get_pg_home_directory
+)
+from . import errors as pgerror
+
+TCP_KEEPIDLE = 24
+TCP_KEEPINTVL = 2
+TCP_KEEPCNT = 3
+
+
+class ConnectionStateType(Enum):
+    CONNECTING = 0
+    SSL_CONNECTING = auto()
+    AUTHENTICATING = auto()
+    SYNCHRONIZING = auto()
+    READY = auto()
+
+
+class Authentication(Enum):
+    NONE = 0
+    PASSWORD = auto()
+    MD5 = auto()
+    SCRAM_SHA256 = auto()
+
+
+@dataclass
+class PGState:
+    parameters: Dict[str, str]
+    cancellation_key: Optional[Tuple[int, int]]
+    auth: Optional[Authentication]
+    server_error: Optional[list[tuple[str, str]]]
+    ssl: bool
+
+
+class ConnectionStateUpdate(Protocol):
+    def send(self, message: memoryview) -> None: ...
+    def upgrade(self) -> None: ...
+    def parameter(self, name: str, value: str) -> None: ...
+    def cancellation_key(self, pid: int, key: int) -> None: ...
+    def state_changed(self, state: int) -> None: ...
+    def auth(self, auth: int) -> None: ...
+
+
+StateChangeCallback = Callable[[ConnectionStateType], None]
+
+
+def _parse_tls_version(tls_version: str) -> ssl_module.TLSVersion:
+    if tls_version.startswith('SSL'):
+        raise ValueError(
+            f"Unsupported TLS version: {tls_version}"
+        )
+    try:
+        return ssl_module.TLSVersion[tls_version.replace('.', '_')]
+    except KeyError:
+        raise ValueError(
+            f"No such TLS version: {tls_version}"
+        )
+
+
+def _create_ssl(ssl_config: Dict[str, Any]):
+    sslmode = SSLMode.parse(ssl_config['sslmode'])
+    ssl = ssl_module.SSLContext(ssl_module.PROTOCOL_TLS_CLIENT)
+    ssl.check_hostname = sslmode >= SSLMode.verify_full
+    if sslmode < SSLMode.require:
+        ssl.verify_mode = ssl_module.CERT_NONE
+    else:
+        if ssl_config['sslrootcert']:
+            ssl.load_verify_locations(ssl_config['sslrootcert'])
+            ssl.verify_mode = ssl_module.CERT_REQUIRED
+        else:
+            if sslmode == SSLMode.require:
+                ssl.verify_mode = ssl_module.CERT_NONE
+        if ssl_config['sslcrl']:
+            ssl.load_verify_locations(ssl_config['sslcrl'])
+            ssl.verify_flags |= ssl_module.VERIFY_CRL_CHECK_CHAIN
+    if ssl_config['sslkey'] and ssl_config['sslcert']:
+        ssl.load_cert_chain(ssl_config['sslcert'],
+                            ssl_config['sslkey'],
+                            ssl_config['sslpassword'] or '')
+    if ssl_config['ssl_max_protocol_version']:
+        ssl.maximum_version = _parse_tls_version(
+            ssl_config['ssl_max_protocol_version'])
+    if ssl_config['ssl_min_protocol_version']:
+        ssl.minimum_version = _parse_tls_version(
+            ssl_config['ssl_min_protocol_version'])
+    # OpenSSL 1.1.1 keylog file
+    if hasattr(ssl, 'keylog_filename'):
+        if ssl_config['keylog_filename']:
+            ssl.keylog_filename = ssl_config['keylog_filename']
+    return ssl
+
+
+class PGConnectionProtocol(asyncio.Protocol):
```

Comment:
Let's compare the performance with `asyncio.BufferedProtocol`, which advocates for zero-copy.

Reply:
I think it might be best to keep it as asyncio.Protocol. This code will go away at some point in the future and I don't believe we were using it before.